### DISCUSSION TITLE
Any examples of calling a Lambda function?

### DISCUSSION THREAD

**Discussion by ToddHoff** [soto-project/soto#725] [General]
Hi, I'd like to call a Lambda function from an iOS app. I've been unable to find any examples of how to do this. Can someone please point me to some example code?

thanks

**Comment by adam-fowler** [comment]
You can call a lambda function directly using the invoke function from the lambda library in Soto. But you are probably better to setup an APIGateway so you can invoke the lambda function with an HTTP client.

There are examples out there to do this. @sebsto do you know where the most up to date guides or examples are for this?

**Comment by sebsto** [comment]
Hello @ToddHoff 
As Adam explained, the most simple way to call a Lambda function is to use the Lambda `InvokeAPI` in Soto (or any AWS SDK, or the AWS CLI). This call requires authentication, you need to have an access key and secret access key to do so.

In the context of an iOS app, this is not the right way to do, because you don't want to expose AWS credentials in your app. 

The most common way to expose a Lambda function to another application is through the API Gateway. You deploy your Lambda function together with an API that will offer an HTTP (TLS) endpoint, take care of the decoding of the parameters (HTTP headers, query params, request params) and send them as a JSON payload to your Lambda function.

The easiest way to deploy such a group of Lambda function + API Gateway is to use SAM.
You can get started with SAM here 
https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/what-is-sam-overview.html

If you write your Lambda functions in Swift, I wrote a tutorial showing you[ how to get started and use SAM to deploy and test your function end to end](https://swiftpackageindex.com/swift-server/swift-aws-lambda-runtime/1.0.0-alpha.2/tutorials/table-of-content).

You can also use the VSCode Lambda SAM plugin to get started with a project with just a few clicks 
https://github.com/swift-server-community/vscode-aws-lambda-swift-sam

These tutorial typically do not address the authentication part.  The best practice here is to protect the API Gateway with a JWT token based authentication and use the authentication token your app received when the user authenticated.
[The API Gateway JWT authorizer doc](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-jwt-authorizer.html) has the details. And [you can configure JWT Authorizers with SAM](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-controlling-access-to-apis-oauth2-authorizer.html) as well

I hope these resources will help you to get started. Let us know if anything is unclear and give us feedback to make them even more useful.

**Comment by sebsto** [comment]
Oh, I should have added, check the [Serverless land pattern collection](https://serverlessland.com/patterns). This is an incredible collection of 830+ serverless patterns, many of them include AWS Lambda, API Gateway and SAM.

**Comment by ToddHoff** [comment]
Thanks. I got the basics from the lambda test harness, which worked great. I appreciate all the hard work that went into the library. Amazing.

I've used the JWT token approach on the web with Cognito. Authentication is by far the hardest part of this whole process, so code examples would be most helpful. 

For this, I think using keys for a user with minimum permissions should work well enough.

**Comment by sebsto** [comment]
I didn't want to mention / push for Cognito in my first reponse to not look like too salesly :-) 
But now that you mentioned it and tried it, the easier way to integrate Cognito, API Gateway, Lambda, and even S3 and DynamoDB from a mobile app is to use the Amplify library.

Amplify allows you to provision these backend resources in a very very easy way, with most of the configuration done for you, and it provides a Swift native client side library to easily access the cloud resources.

I contributed to the [Amplify Swift getting started guide](https://docs.amplify.aws/swift/) (be sure to select Gen - read it, give it a try and [give us feedback](https://github.com/aws-amplify/amplify-swift).

**Reply by ToddHoff** [reply]
Interesting, thanks. I'll take a look
